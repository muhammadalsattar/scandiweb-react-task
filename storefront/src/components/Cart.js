import React from 'react';
import { connect } from 'react-redux';
import Navbar from './Navbar';
import {
    pickPrice,
    calculateTotal,
    calculateQuantity,
    calcualteTax,
} from '../utils/utils';
import { removeFromCart, addToCart } from '../actions/cart';

export class Cart extends React.Component {
    componentDidMount() {
        document
            .querySelectorAll('.cart .attribute-value button')
            .forEach(button => {
                button.style.backgroundColor = button.id;
            });
    }
    componentDidUpdate() {
        document
            .querySelectorAll('.cart .attribute-value button')
            .forEach(button => {
                button.style.backgroundColor = button.id;
            });
    }
    updateCart = product => {
        this.props.updateCart(product);
    };
    removeItem = product => {
        this.props.removeItem(product);
    };
    changeImage = e => {
        e.preventDefault();
        const gallery = this.props.cart.find(
            product => product.id === e.target.id
        ).gallery;
        const index = gallery.indexOf(
            e.target.parentNode.parentNode.firstChild.src
        );
        if (e.target.innerHTML === '&lt;') {
            if (index > 0) {
                e.target.parentNode.parentNode.firstChild.src =
                    gallery[index - 1];
            } else {
                e.target.parentNode.parentNode.firstChild.src =
                    gallery[gallery.length - 1];
            }
        } else {
            if (index < gallery.length - 1) {
                e.target.parentNode.parentNode.firstChild.src =
                    gallery[index + 1];
            } else {
                e.target.parentNode.parentNode.firstChild.src = gallery[0];
            }
        }
    };
    render() {
        return (
            <div className="cart-page">
                <Navbar />
                <div className="cart page-body">
                    <h1 className="cart-heading">Cart</h1>
                    <div className="cart-items">
                        {this.props.cart &&
                            this.props.cart.map(
                                (
                                    {
                                        id,
                                        brand,
                                        name,
                                        prices,
                                        attributes,
                                        selectedAttr,
                                        quantity,
                                        gallery,
                                    },
                                    index
                                ) => (
                                    <div className="cart-item" key={index}>
                                        <div className="item-details">
                                            <h2 className="brand">{brand}</h2>
                                            <h2 className="name">{name}</h2>
                                            <h3 className="price">
                                                {
                                                    this.props.defaultCurrency
                                                        .symbol
                                                }
                                                {pickPrice(
                                                    prices,
                                                    this.props.defaultCurrency
                                                )}
                                            </h3>
                                            {attributes.map(
                                                ({ name, items }, index) => (
                                                    <div
                                                        className="cart-item-attribute"
                                                        key={name}
                                                    >
                                                        <p className="attribute-name">
                                                            {name}:
                                                        </p>
                                                        <div className="attribute-value">
                                                            {items.map(
                                                                ({ value }) =>
                                                                    name ===
                                                                    'Color' ? (
                                                                        value ===
                                                                        selectedAttr[
                                                                            index
                                                                        ]
                                                                            .value ? (
                                                                            <div
                                                                                className={
                                                                                    name +
                                                                                    ' selected'
                                                                                }
                                                                                key={
                                                                                    value
                                                                                }
                                                                            >
                                                                                <button
                                                                                    id={
                                                                                        value
                                                                                    }
                                                                                />
                                                                            </div>
                                                                        ) : (
                                                                            <div
                                                                                className={
                                                                                    name
                                                                                }
                                                                                key={
                                                                                    value
                                                                                }
                                                                            >
                                                                                <button
                                                                                    id={
                                                                                        value
                                                                                    }
                                                                                />
                                                                            </div>
                                                                        )
                                                                    ) : value ===
                                                                      selectedAttr[
                                                                          index
                                                                      ]
                                                                          .value ? (
                                                                        <div
                                                                            className="other-attr selected"
                                                                            key={
                                                                                value
                                                                            }
                                                                        >
                                                                            <button>
                                                                                {
                                                                                    value
                                                                                }
                                                                            </button>
                                                                        </div>
                                                                    ) : (
                                                                        <div
                                                                            className="other-attr"
                                                                            key={
                                                                                value
                                                                            }
                                                                        >
                                                                            <button>
                                                                                {
                                                                                    value
                                                                                }
                                                                            </button>
                                                                        </div>
                                                                    )
                                                            )}
                                                        </div>
                                                    </div>
                                                )
                                            )}
                                        </div>
                                        <div className="quantity-image">
                                            <div className="quantity">
                                                <button
                                                    onClick={() => {
                                                        this.updateCart({
                                                            id,
                                                            quantity: 1,
                                                            selectedAttr,
                                                        });
                                                    }}
                                                >
                                                    +
                                                </button>
                                                <p>{quantity}</p>
                                                <button
                                                    onClick={() => {
                                                        quantity > 1
                                                            ? this.updateCart({
                                                                  id,
                                                                  quantity: -1,
                                                                  selectedAttr,
                                                              })
                                                            : this.removeItem({
                                                                  id,
                                                                  selectedAttr,
                                                              });
                                                    }}
                                                >
                                                    -
                                                </button>
                                            </div>
                                            <div className="image">
                                                <img
                                                    src={gallery[0]}
                                                    alt={name}
                                                ></img>
                                                {gallery.length > 1 && (
                                                    <div className="browse">
                                                        <button
                                                            id={id}
                                                            onClick={
                                                                this.changeImage
                                                            }
                                                        >{`<`}</button>
                                                        <button
                                                            id={id}
                                                            onClick={
                                                                this.changeImage
                                                            }
                                                        >{`>`}</button>
                                                    </div>
                                                )}
                                            </div>
                                        </div>
                                    </div>
                                )
                            )}
                    </div>
                    <div className="cart-summary">
                        <p>
                            Tax 21%:{' '}
                            <b>
                                {this.props.defaultCurrency?.symbol}{' '}
                                {calcualteTax(
                                    calculateTotal(
                                        this.props.cart,
                                        this.props.defaultCurrency
                                    )
                                )}
                            </b>
                        </p>
                        <p>
                            Quantity:{' '}
                            <b>{calculateQuantity(this.props.cart)}</b>
                        </p>
                        <p>
                            Total:{' '}
                            <b>
                                {this.props.defaultCurrency?.symbol}{' '}
                                {calculateTotal(
                                    this.props.cart,
                                    this.props.defaultCurrency
                                )}
                            </b>
                        </p>
                        <button>Order</button>
                    </div>
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = dispatch => {
    return {
        updateCart: product => {
            dispatch(addToCart(product));
        },
        removeItem: product => {
            dispatch(removeFromCart(product));
        },
    };
};

const mapStateToProps = ({ cart, currencies }) => {
    return {
        cart: cart.cart,
        defaultCurrency: currencies.defaultCurrency,
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(Cart);
