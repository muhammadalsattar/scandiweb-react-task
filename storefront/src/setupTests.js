const currencies = [
    {
        label: 'USD',
        symbol: '$',
    },
    {
        label: 'GBP',
        symbol: 'Â£',
    },
];
const categories = [
    {
        name: 'category1',
    },
    {
        name: 'category2',
    },
];
const attributes = [
    {
        id: 'Size',
        name: 'Size',
        type: 'text',
        items: [
            {
                id: 40,
                value: '40',
                displayValue: '40',
            },
        ],
    },
    {
        id: 'Color',
        name: 'Color',
        type: 'swatch',
        items: [
            {
                id: 'red',
                value: 'red',
                displayValue: 'Red',
            },
        ],
    },
];
const gallery = [
    'https://dummyimage.com/600x400/000/fff',
    'https://dummyimage.com/600x400/000/ggg',
    'https://dummyimage.com/600x400/000/hhh',
];
const prices = [
    {
        amount: 518.47,
        currency: currencies[0],
    },
    {
        amount: 372.67,
        currency: currencies[1],
    },
];
const products = [
    {
        id: '1',
        name: 'product1',
        brand: 'brand1',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[0],
        inStock: true,
        quantity: 1,
        selectedAttr: [{ name: 'Size', value: '40' }],
    },
    {
        id: '2',
        name: 'product2',
        brand: 'brand2',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[1],
        inStock: false,
        quantity: 1,
        selectedAttr: [{ name: 'Color', value: 'red' }],
    },
    {
        id: '3',
        name: 'product3',
        brand: 'brand3',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[0],
        inStock: false,
        quantity: 1,
        selectedAttr: [{ name: 'Size', value: '41' }],
    },
    {
        id: '4',
        name: 'product4',
        brand: 'brand4',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[1],
        inStock: false,
        quantity: 1,
        selectedAttr: [{ name: 'Color', value: 'orange' }],
    },
    {
        id: '5',
        name: 'product5',
        brand: 'brand5',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[0],
        inStock: true,
        quantity: 1,
        selectedAttr: [{ name: 'Size', value: '42' }],
    },
    {
        id: '6',
        name: 'product6',
        brand: 'brand6',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[1],
        inStock: false,
        quantity: 1,
        selectedAttr: [{ name: 'Color', value: 'green' }],
    },
];

export const initialState = {
    currencies,
    categories,
    products,
};
