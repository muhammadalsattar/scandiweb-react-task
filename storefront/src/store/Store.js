import { configureStore, combineReducers } from '@reduxjs/toolkit';
import cartReducer from '../reducers/cartReducer';
import productsReducer from '../reducers/productsReducer';
import categoriesReducer from '../reducers/categoriesReducer';
import currencyReducer from '../reducers/currencyReducer';

const rootReducer = combineReducers({
    cart: cartReducer,
    products: productsReducer,
    categories: categoriesReducer,
    currencies: currencyReducer,
});
// creating store and combining reducers
const Store = configureStore({
    reducer: rootReducer,
});

export default Store;
