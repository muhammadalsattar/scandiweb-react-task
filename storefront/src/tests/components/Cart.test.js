import React from "react";
import {render, screen} from "@testing-library/react";
import { Provider } from "react-redux";
import {BrowserRouter as Router} from "react-router-dom";
import Store from "../../store/Store.js";
import {Cart} from "../../components/Cart.js";
import { initialState } from "../../setupTests.js"

test("Cart renders correctly", () => {
    const { undefined, container } = render(
        <Provider store={Store}>
        <Router><Cart cart={initialState.cart} defaultCurrency={initialState.currencies[0]}/></Router>
        </Provider>
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByText(/1178.01/i)).toBeInTheDocument();
});