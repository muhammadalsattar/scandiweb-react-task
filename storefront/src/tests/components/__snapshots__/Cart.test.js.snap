// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart renders correctly 1`] = `
<div>
  <div
    class="cart-page"
  >
    <nav
      class="navbar"
    >
      <div
        class="categories"
      />
      <div
        class="brand-logo"
      >
        <a
          href="/"
        >
          <img
            alt="logo-icon"
            src="/logo.png"
          />
        </a>
      </div>
      <div
        class="dropdowns"
      >
        <div
          class="currency-div"
        >
          <button
            class="currency-icon"
          >
            <span
              class="currency-drop-icon"
            >
              âˆ§
            </span>
          </button>
          <ul
            class="currency-dropdown"
            hidden=""
          />
        </div>
        <div
          class="cart-div"
        >
          <button
            class="cart-img"
          >
            <img
              alt="cart-img"
              src="/cart.png"
            />
          </button>
          <button
            class="cart-count"
          >
            0
          </button>
          <div
            class="cart-overlay"
            hidden=""
          >
            <p>
              <b>
                My Bag, 
              </b>
              0
               items
            </p>
            <div
              class="cart-total"
            >
              <p>
                Total
              </p>
              <p>
                0.00
              </p>
            </div>
            <div
              class="bag-checkout"
            >
              <div
                class="view-bag"
              >
                <a
                  href="/cart"
                >
                  <button>
                    View Bag
                  </button>
                </a>
              </div>
              <div
                class="checkout"
              >
                <button>
                  Check out
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </nav>
    <div
      class="cart page-body"
    >
      <h1
        class="cart-heading"
      >
        Cart
      </h1>
      <div
        class="cart-items"
      >
        <div
          class="cart-item"
        >
          <div
            class="item-details"
          >
            <h2
              class="brand"
            >
              Sony
            </h2>
            <h2
              class="name"
            >
              PlayStation 5
            </h2>
            <h3
              class="price"
            >
              $
              844.02
            </h3>
            <div
              class="cart-item-attribute"
            >
              <p
                class="attribute-name"
              >
                Color
                :
              </p>
              <div
                class="attribute-value"
              >
                <div
                  class="Color selected"
                >
                  <button
                    style="background-color: rgb(68, 255, 3);"
                  />
                </div>
                <div
                  class="Color"
                >
                  <button
                    style="background-color: rgb(3, 255, 247);"
                  />
                </div>
                <div
                  class="Color"
                >
                  <button
                    style="background-color: rgb(3, 11, 255);"
                  />
                </div>
                <div
                  class="Color"
                >
                  <button
                    style="background-color: rgb(0, 0, 0);"
                  />
                </div>
                <div
                  class="Color"
                >
                  <button
                    style="background-color: rgb(255, 255, 255);"
                  />
                </div>
              </div>
            </div>
            <div
              class="cart-item-attribute"
            >
              <p
                class="attribute-name"
              >
                Capacity
                :
              </p>
              <div
                class="attribute-value"
              >
                <div
                  class="other-attr selected"
                >
                  <button>
                    512G
                  </button>
                </div>
                <div
                  class="other-attr"
                >
                  <button>
                    1T
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            class="quantity-image"
          >
            <div
              class="quantity"
            >
              <button>
                +
              </button>
              <p>
                1
              </p>
              <button>
                -
              </button>
            </div>
            <div
              class="image"
            >
              <img
                alt="PlayStation 5"
                src="https://images-na.ssl-images-amazon.com/images/I/510VSJ9mWDL._SL1262_.jpg"
              />
              <div
                class="browse"
              >
                <button
                  id="ps-5"
                >
                  &lt;
                </button>
                <button
                  id="ps-5"
                >
                  &gt;
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="cart-item"
        >
          <div
            class="item-details"
          >
            <h2
              class="brand"
            >
              Microsoft
            </h2>
            <h2
              class="name"
            >
              Xbox Series S 512GB
            </h2>
            <h3
              class="price"
            >
              $
              333.99
            </h3>
            <div
              class="cart-item-attribute"
            >
              <p
                class="attribute-name"
              >
                Color
                :
              </p>
              <div
                class="attribute-value"
              >
                <div
                  class="Color selected"
                >
                  <button
                    style="background-color: rgb(68, 255, 3);"
                  />
                </div>
                <div
                  class="Color"
                >
                  <button
                    style="background-color: rgb(3, 255, 247);"
                  />
                </div>
                <div
                  class="Color"
                >
                  <button
                    style="background-color: rgb(3, 11, 255);"
                  />
                </div>
                <div
                  class="Color"
                >
                  <button
                    style="background-color: rgb(0, 0, 0);"
                  />
                </div>
                <div
                  class="Color"
                >
                  <button
                    style="background-color: rgb(255, 255, 255);"
                  />
                </div>
              </div>
            </div>
            <div
              class="cart-item-attribute"
            >
              <p
                class="attribute-name"
              >
                Capacity
                :
              </p>
              <div
                class="attribute-value"
              >
                <div
                  class="other-attr selected"
                >
                  <button>
                    512G
                  </button>
                </div>
                <div
                  class="other-attr"
                >
                  <button>
                    1T
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            class="quantity-image"
          >
            <div
              class="quantity"
            >
              <button>
                +
              </button>
              <p>
                1
              </p>
              <button>
                -
              </button>
            </div>
            <div
              class="image"
            >
              <img
                alt="Xbox Series S 512GB"
                src="https://images-na.ssl-images-amazon.com/images/I/71vPCX0bS-L._SL1500_.jpg"
              />
              <div
                class="browse"
              >
                <button
                  id="xbox-series-s"
                >
                  &lt;
                </button>
                <button
                  id="xbox-series-s"
                >
                  &gt;
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="cart-summary"
      >
        <p>
          Tax 21%: 
          <b>
            $
             
            247.38
          </b>
        </p>
        <p>
          Quantity: 
          <b>
            2
          </b>
        </p>
        <p>
          Total: 
          <b>
            $
             
            1178.01
          </b>
        </p>
        <button>
          Order
        </button>
      </div>
    </div>
  </div>
</div>
`;
